crates/dbsp/src/operator/dynamic/aggregate/chain_aggregate.rs

```diff
@@ -0,0 +1,239 @@
+                    }
+                }
+            };
+
+            // If there is an existing value, start from it, otherwise start from finit().
+            if retract {
+                old.clone_to(&mut new);
+            } else {
+                (self.finit)(&mut new);
+            }
+
+            // Iterate over new values.
+            while delta_cursor.val_valid() {
+                let w = **delta_cursor.weight();
+                (self.fupdate)(&mut new, delta_cursor.val(), w);
+                delta_cursor.step_val();
+            }
+
+            // Apply retraction and insertion in correct order.
+            if retract {
+                match new.cmp(&old) {
+                    Ordering::Less => {
+                        builder.push_refs(&key, &new, (1 as ZWeight).erase());
+                        builder.push_vals(&mut key, &mut old, (-1 as ZWeight).erase_mut());
+                    }
+                    Ordering::Greater => {
+                        builder.push_refs(&key, &old, (-1 as ZWeight).erase());
+                        builder.push_vals(&mut key, &mut new, (1 as ZWeight).erase_mut());
+                    }
+                    _ => (),
+                }
+                // do nothing if new == old.
+            } else {
+                builder.push_vals(&mut key, &mut new, 1.erase_mut());
+            }
+
+            delta_cursor.step_key();
+        }
+        builder.done()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{cmp::min, i32};
+
+    use crate::{
+        circuit::CircuitConfig, operator::Min, utils::Tup2, OrdIndexedZSet, OutputHandle,
+        RootCircuit, Runtime, ZSetHandle, ZWeight,
+    };
+    use proptest::{collection, prelude::*};
+
+    fn min_test_circuit(
+        circuit: &mut RootCircuit,
+    ) -> (
+        ZSetHandle<Tup2<u64, i32>>,
+        OutputHandle<OrdIndexedZSet<u64, i32>>,
+        OutputHandle<OrdIndexedZSet<u64, i32>>,
+    ) {
+        let (input, input_handle) = circuit.add_input_zset::<Tup2<u64, i32>>();
+
+        let input_indexed = input.map_index(|Tup2(x, y)| (*x, *y));
+        let aggregate = input_indexed.aggregate(Min).output();
+        let delta_aggregate = input_indexed
+            .chain_aggregate(|| i32::MAX, |acc, i, _w| *acc = min(*acc, *i))
+            .output();
+
+        (input_handle, aggregate, delta_aggregate)
+    }
+
+    pub fn test_input(
+        num_keys: u64,
+        max_val: i32,
+        max_tuples: usize,
+    ) -> impl Strategy<Value = Vec<Tup2<Tup2<u64, i32>, ZWeight>>> {
+        collection::vec(
+            (
+                (0..num_keys, -max_val..max_val).prop_map(|(x, y)| Tup2(x, y)),
+                1..=2i64,
+            )
+                .prop_map(|(x, y)| Tup2(x, y)),
+            0..max_tuples,
+        )
+    }
+
+    pub fn test_inputs() -> impl Strategy<Value = Vec<Vec<Tup2<Tup2<u64, i32>, ZWeight>>>> {
+        collection::vec(test_input(20, 1000, 30), 0..10)
+    }
+
+    proptest! {
```

Comment:
does this work for arg_min as well?
a test would prove it

Reply:
Because of how this operator works, arg_min can only be implemented by attaching a map to its output. I'll add a test like that.